### ISSUE TITLE
unexpected eager expansion of union of sealed subclasses in typechecking

### ISSUE THREAD

**Issue by froydnj** [sorbet/sorbet#8357]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20%3A%3AModule%0A%20%20include%20T%3A%3ASig%0Aend%0A%0Amodule%20Payment%0A%20%20extend%20T%3A%3AHelpers%0A%20%20sealed!%0A%20%20abstract!%0A%0A%20%20sig%20%7Babstract.returns%28T.any%28String%2C%20Integer%29%29%7D%0A%20%20def%20amount%3B%20end%0A%0A%20%20class%20Canceled%20%3C%20T%3A%3AStruct%0A%20%20%20%20include%20Payment%0A%0A%20%20%20%20const%20%3Aamount%2C%20T.any%28String%2C%20Integer%29%0A%20%20end%0A%0A%20%20class%20Open%20%3C%20T%3A%3AStruct%0A%20%20%20%20include%20Payment%0A%0A%20%20%20%20const%20%3Aamount%2C%20T.any%28String%2C%20Integer%29%0A%20%20end%0A%0A%20%20class%20Pending%20%3C%20T%3A%3AStruct%0A%20%20%20%20include%20Payment%0A%0A%20%20%20%20const%20%3Aamount%2C%20T.any%28String%2C%20Integer%2C%20Float%29%0A%20%20end%0Aend%0A%0Asig%20%7Bparams%28p%3A%20T.nilable%28Payment%29%29.void%7D%0Adef%20handle_payment%28p%29%0A%20%20return%20unless%20p%0A%0A%20%20T.reveal_type%28p.amount%29%0Aend)

```ruby
# typed: true

class ::Module
  include T::Sig
end

module Payment
  extend T::Helpers
  sealed!
  abstract!

  sig {abstract.returns(T.any(String, Integer))}
  def amount; end

  class Canceled < T::Struct
    include Payment

    const :amount, T.any(String, Integer)
  end

  class Open < T::Struct
    include Payment

    const :amount, T.any(String, Integer)
  end

  class Pending < T::Struct
    include Payment

    const :amount, T.any(String, Integer, Float)
  end
end

sig {params(p: T.nilable(Payment)).void}
def handle_payment(p)
  return unless p

  T.reveal_type(p.amount)
end
```

#### Observed output

```
editor.rb:38: Revealed type: T.any(Float, String, Integer) https://srb.help/7014
    38 |  T.reveal_type(p.amount)
          ^^^^^^^^^^^^^^^^^^^^^^^
  Got T.any(Float, String, Integer) originating from:
    editor.rb:38:
    38 |  T.reveal_type(p.amount)
                        ^^^^^^^^
Errors: 1
```

#### Expected behavior

This cut-down example comes from a user, who was surprised to see that they needed to handle three different types when the module interface they were accepting only requires two.  `p` is of type `T.any(Payment::Open, Payment::Pending, Payment::Canceled)`, not `Payment`.

There is some overlap here with #4100 -- we shouldn't have permitted the `Pending#amount` definition in the first place.  (We get errors if we declare a proper `override` function with a `T.any(String, Integer, Float)` return type.)

In some ways, the eager expansion here is a good thing -- Sorbet is saving the user from potential problems down the line.  But it's not terribly intuitive, IMHO.
- - -

**Comment by jez** [comment]
> In some ways, the eager expansion here is a good thing

Yes, in so far as it happens to be a safety net against certain kinds of missing override checking, like you mentioned.

But absent that, Sorbet should be able to treat a sealed class as a union of its subclasses without loss in precision—in an ideal world, the only reason why Sorbet chooses not to expand something is for performance, not correctness.

So technically, Sorbet could drop the `nil` from `T.nilable(Payment)` without needing to expand it, as a performance optimization (at the current expense of making code more susceptible to override checking bugs).
